### DISCUSSION TITLE
Neovim LSP Autocomplete

### DISCUSSION THREAD

**Discussion by joshmedeski** [Hammerspoon/hammerspoon#3451] [Ideas]
I'm using Neovim to write my hammerspoon configuration.

I'd love to setup auto-complete, code completion, and documentation for the lua language server.

What would be the first steps to getting this done? Has anyone else made any attempts to work on this?

**Comment by megalithic** [comment]
Hey there @joshmedeski; the best bet is to first, install the `EmmyLua` spoon. Then, you can include that path in the ones `lua_ls` uses; I've done something like that here: https://github.com/megalithic/dotfiles/blob/main/config/nvim/lua/mega/servers.lua#L404

![clipboard.png](https://i.imgur.com/RNDRJfg.jpeg)

**Comment by joshmedeski** [comment]
Thanks, once I installed and loaded the EmmyLua Spoon, I created this `.neoconf.json` in my hammerspoon config directory.

```json
{
  "neodev": {
    "library": {
      "enabled": true,
      "plugins": ["nvim-lspconfig"]
    }
  },
  "lspconfig": {
    "lua_ls": {
      "Lua.workspace.library": [
        "/Users/{username}/.hammerspoon/Spoons/EmmyLua.spoon/annotations"
      ]
    }
  }
}
```

And it works!

**Reply by megalithic** [reply]
brilliant!

**Reply by Demianeen** [reply]
Just for everybody else as stupid as me. You need to load Spoon first!

**Comment by tsxoxo** [comment]
Thank You for this thread! This still works in 2025. I always feel a little pity for my LSP when it does not know what is going on.

## Step-by-step instructions for noobs like me

*Assumes you have hammerspoon installed.*

* Download the [EmmyLua spoon](https://www.hammerspoon.org/Spoons/EmmyLua.html)
* Load it:

```lua
-- {your hammerspoon path}/init.lua
hs.loadSpoon("EmmyLua")
```

* Reload config. EmmyLua will generate the annotations your LSP needs (I think)
* Point your LSP to the annotations:

Setup for NeoVim 0.11 with native LSP

```lua
-- nvim/lsp/lua_ls.lua

return {
	on_init = function(client)
		client.config.settings.Lua = vim.tbl_deep_extend("force", client.config.settings.Lua, {
			workspace = {
				library = {
					vim.fn.expand("~/.hammerspoon/Spoons/EmmyLua.spoon/annotations"),
				},
			},
		})
	end,
}
```

---

That's how I got it to work, anyway. 

Any optimizations welcome!

Cheers!
tsxoxo

PS. As long as I don't install new Spoons, I don't need to load EmmyLua each time, right?
